gcc/rust/expand/rust-derive-eq.cc

```diff
@@ -0,0 +1,213 @@
+      Visibility::create_private (), loc)});
+}
+
+std::unique_ptr<Stmt>
+DeriveEq::assert_type_is_eq (std::unique_ptr<Type> &&type)
+{
+  auto assert_param_is_eq = "AssertParamIsEq";
+
+  // AssertParamIsCopy::<Self>
+  auto assert_param_is_eq_ty
+    = std::unique_ptr<TypePathSegment> (new TypePathSegmentGeneric (
+      PathIdentSegment (assert_param_is_eq, loc), false,
+      GenericArgs ({}, {GenericArg::create_type (std::move (type))}, {}, loc),
+      loc));
+
+  // TODO: Improve this, it's really ugly
+  auto type_paths = std::vector<std::unique_ptr<TypePathSegment>> ();
+  type_paths.emplace_back (std::move (assert_param_is_eq_ty));
+
+  auto full_path
+    = std::unique_ptr<Type> (new TypePath ({std::move (type_paths)}, loc));
+
+  return builder.let (builder.wildcard (), std::move (full_path));
+}
+
+std::vector<std::unique_ptr<Item>>
+DeriveEq::eq_impls (
+  std::unique_ptr<AssociatedItem> &&fn, std::string name,
+  const std::vector<std::unique_ptr<GenericParam>> &type_generics)
+{
+  auto eq = builder.type_path ({"core", "cmp", "Eq"}, true);
+  auto steq = builder.type_path (LangItem::Kind::STRUCTURAL_TEQ);
+
+  auto trait_items = vec (std::move (fn));
+
+  auto eq_generics
+    = setup_impl_generics (name, type_generics, builder.trait_bound (eq));
+  auto steq_generics = setup_impl_generics (name, type_generics);
+
+  auto eq_impl = builder.trait_impl (eq, std::move (eq_generics.self_type),
+				     std::move (trait_items),
+				     std::move (eq_generics.impl));
+  auto steq_impl
+    = builder.trait_impl (steq, std::move (steq_generics.self_type),
+			  std::move (trait_items),
+			  std::move (steq_generics.impl));
+
+  return vec (std::move (eq_impl), std::move (steq_impl));
+}
+
+void
+DeriveEq::visit_tuple (TupleStruct &item)
+{
+  auto types = std::vector<std::unique_ptr<Type>> ();
+
+  for (auto &field : item.get_fields ())
+    types.emplace_back (field.get_field_type ().clone_type ());
+
+  expanded = eq_impls (assert_receiver_is_total_eq_fn (std::move (types)),
+		       item.get_identifier ().as_string (),
+		       item.get_generic_params ());
+}
+
+void
+DeriveEq::visit_struct (StructStruct &item)
+{
+  auto types = std::vector<std::unique_ptr<Type>> ();
+
+  for (auto &field : item.get_fields ())
+    types.emplace_back (field.get_field_type ().clone_type ());
+
+  expanded = eq_impls (assert_receiver_is_total_eq_fn (std::move (types)),
+		       item.get_identifier ().as_string (),
+		       item.get_generic_params ());
+}
+
+void
+DeriveEq::visit_enum (Enum &item)
+{
+  auto types = std::vector<std::unique_ptr<Type>> ();
+
+  for (auto &variant : item.get_variants ())
+    {
+      switch (variant->get_enum_item_kind ())
+	{
+	case EnumItem::Kind::Identifier:
+	case EnumItem::Kind::Discriminant:
+	  // nothing to do as they contain no inner types
+	  continue;
+	  case EnumItem::Kind::Tuple: {
+	    auto tuple = static_cast<EnumItemTuple &> (*variant);
```

Comment:
We want to avoid the copy here `auto &tuple = static_cast<EnumItemTuple &> (*variant);`.